### ISSUE TITLE
Server disappears from iOS app for no apparently reason

### ISSUE THREAD

**Issue by jaredgisin** [home-assistant/iOS#3141]
**iOS device model, version and app version**

Model Name: iPhone 13 Pro Max
Software Version:  iOS 18.0.1
App version: 2024.9.4

**Home Assistant Core Version**
2024.11.1

**Describe the bug**

I have a server on my local network that appears in the "Add Server" list. It appears there with a DNS name that works on my local network. I can add the server, authenticate, login just fine. The server shows up in the list with the cloud icon next to it. However at some point after, not sure what triggers this, the server just disappears from the list.

**To Reproduce**
See steps above. 

**Expected behavior**
Once added, a server should NEVER disappear from the list of servers that I have added. It's a horrific UX to have items that the users adds to a list simply disappear for now logical reason. If there's problem communicating with the server, authentication issue, network problem, whatever, the server should stay in the list and it should indicate that it's in an error state and the UI should provide a way for the user to correct the problem. It should NOT just disappear from the list for ANY reason whatsoever. Any server that was valid when it's add should persist in that list even if it later becomes invalid for any reason.

**Screenshots**

**Additional context**

**Comment by bgoncal** [comment]
Any chances that your instance returns "unauthorized" at any point? If the app receives unauthorized from the host it reopens the onboarding flow and asks to login again.

**Comment by guidau** [comment]
I have had the same problem for about 4 days. I use Nabu Casa and a virtual HAOS image. I have the same behavior on my iPhone and iPad.

**Comment by jaredgisin** [comment]
> Any chances that your instance returns "unauthorized" at any point? If the app receives unauthorized from the host it reopens the onboarding flow and asks to login again.

I don't actually know, but I highly doubt it. I am able to log in the first time just fine when the server is added. The server stays in the app and functions normally, and then at some point it just disappears.

Even if it were getting "unauthorized" at any point, at no point do I ever see the onboarding flow reopened. I am never prompted to re-authenticate. If that were the case then I wouldn't have opened this issue.

To be clear, this issue is about the fact that the iOS app DROPS the server from the list. It flat out DISAPPEARS from the list of server which is a significant  usability problem because it doesn't give me any change to correct issues with it. I argue that the correct usability is that if I add a server to the list, and I am able to authenticate the server the first time to have it then appear in the list, there should be never be any circumstance in which the iOS app decides to drop the server from that list. If it starts unauthenticating, or being unreachable or whatever the hell DO NOT DROP THE SERVER FROM THE LIST. The user should see the server in the list at all time, be notified there is an issue with it and given the ability to reuathentica it or restore the connectivity to it, or delete themselves.

So the question here and the issue that needs to be resolved is to figure out under what circumstances the iOS app simply just deletes the server record from the servers list and to stop that from happening. Secondarily, then we can figure out if there is some connectivity or authentication issue with my setup. I would be far less upset about this and more likely to resolve networking or authentication issue if the server were to remain int he list and to show me an error about why the app is unable to connect, but as it stands now as the server is flat out dropped with zero explanation, I don't even know where to start to figure out if I have an issue.

**Comment by bgoncal** [comment]
Do you have more than 1 server? Because if the server vanishes from the list and there are no other servers, you should see the onboarding again. can you share a screen recording when that happens again?

**Comment by guidau** [comment]
It seems I have solved my problem! In my case, the unauthorized access was triggered via Nabu Casa, the local IP 127.0.0.1 was blocked. 
I have found the following help for this: [https://nabucasa.com/config/troubleshooting/403_forbidden/]
However, I still can't understand why the server entry in the configuration is simply deleted in the event of access problems!

**Comment by bgoncal** [comment]
Message received that this is a bad UX, now we need to find out what is triggering that, as my initial theory at least for @guidau the issue was the unauthorized response

**Comment by bgoncal** [comment]
@ChristopherGerdes yes you can find out what is happening because if you export your companion app logs there should be something there to help.
Feel free to submit here as well https://forms.gle/Uoqz127Phx4mMTpS6

**Comment by ChristopherGerdes** [comment]
> It seems I have solved my problem! In my case, the unauthorized access was triggered via Nabu Casa, the local IP 127.0.0.1 was blocked. I have found the following help for this: [https://nabucasa.com/config/troubleshooting/403_forbidden/] However, I still can't understand why the server entry in the configuration is simply deleted in the event of access problems!

ok, I may be encountering this.  Did you also need to add anything to your config to prevent it from being banned in the future?  I don't see anything in the instructions about that, but it seems like it would need to be added to trusted proxies or something to keep it from happening again.

**Comment by bgoncal** [comment]
@guidau in case you receive "unauthorized" and you are lead to the onboarding, if you sign in again the server will return, having the server in the list but not working is also useless right? 

Again, the UX CAN and WILL be improved, I'm just brainstorming with you so we find a nice approach

**Comment by ChristopherGerdes** [comment]
> @guidau in case you receive "unauthorized" and you are lead to the onboarding, if you sign in again the server will return, having the server in the list but not working is also useless right?
> 
> Again, the UX CAN and WILL be improved, I'm just brainstorming with you so we find a nice approach

Thank you for this.  You may be hearing frustration from us simply because I think a number of us have been facing this for a while and it's like slamming your head against the wall.  I know I definitely appreciate the help.

**Comment by bgoncal** [comment]
@ChristopherGerdes after the server gets deleted, add it again then go to companion app settings >> debugging >> export logs

**Comment by ChristopherGerdes** [comment]
> @ChristopherGerdes yes you can find out what is happening because if you export your companion app logs there should be something there to help. Feel free to submit here as well https://forms.gle/Uoqz127Phx4mMTpS6

Thank you for this!  I've grabbed the logs from my ipad and submitted them.

**Comment by JManch** [comment]
For anyone running into this issue whilst hosting Home Assistant behind their own reverse proxy, I was able to workaround it by responding with a 404 instead of a 403 in the case of unauthorized access.

https://github.com/home-assistant/iOS/blob/8c30c7cd85c72fe3807bb8bc0c686975625cb1f7/Sources/Shared/API/Authentication/TokenManager.swift#L155-L167

**Comment by jaredgisin** [comment]
> Do you have more than 1 server? Because if the server vanishes from the list and there are no other servers, you should see the onboarding again. can you share a screen recording when that happens again?

Yes, I have more than one server. Two of them never go away. They are running the HAOS/supervisor setup on RaspberryPI and I have them added using the IP address of each host. They never disappear from the iOS app. 

There is no way to do any screen recording here as other have stated the server simply just disappear at some point. Likely ti ay be when I leave the local network, but this seems strange to as when they are connected locally they show the cloud icon next to them and when they are added they show they are using the internal URL, so I would expect them to work when I am not on the local network but should switch to cloud. That doesn't seem to happen, but I can't say for sure that's when/what causes these to disappear as I have never tested.

**Comment by jaredgisin** [comment]
> @guidau in case you receive "unauthorized" and you are lead to the onboarding, if you sign in again the server will return, having the server in the list but not working is also useless right?
> 
> Again, the UX CAN and WILL be improved, I'm just brainstorming with you so we find a nice approach

I would imagine that simple code inspection should reveal all of the places in the code where the server is removed from the list, where the configuration is deleted. As a developer myself, I would look at how that can happen in the code and remove that functionality. That should also point to the reason why it can happen.

**Comment by bgoncal** [comment]
The cloud icon is just a label that shows if you have HA Cloud configured, nothing else.

Between your 3 servers, what's the difference from the 2 that never get deleted to the third one in terms of configuration? Does any of them have a user with local access only? Any kind of security measure that can prevent the remote access?

**Comment by bgoncal** [comment]
It's not "just" remove the code that deletes/hide the server, there are many things and flows that took this into consideration so they all need to be changed as well.
Before moving forward I want to validate if all cases that this happened were due to unauthorized access, till now, all were.

**Comment by bgoncal** [comment]
> For anyone running into this issue whilst hosting Home Assistant behind their own reverse proxy, I was able to workaround it by responding with a 404 instead of a 403 in the case of unauthorized access.
> 
> 
> 
> https://github.com/home-assistant/iOS/blob/8c30c7cd85c72fe3807bb8bc0c686975625cb1f7/Sources/Shared/API/Authentication/TokenManager.swift#L155-L167

Thanks for pointing out

**Comment by jaredgisin** [comment]
> The cloud icon is just a label that shows if you have HA Cloud configured, nothing else.
> 
> Between your 3 servers, what's the difference from the 2 that never get deleted to the third one in terms of configuration? Does any of them have a user with local access only? Any kind of security measure that can prevent the remote access?

The difference  is that 2 that never get deleted have been manually added using the IPv4 address of the Raspberry Pi 4 that they are running on my local network. These are 192.168.x.x addresses. Those never get deleted and are always there.

The one that gets deleted is running as a docker container Nginx reverse proxy on the same host. That system has a local DNS name such as ha.blah.network that my local Unifi device serves DNS for. That DNS name is then configured as a host the Nginx reverse proxy which directs traffic over the local docker network within that host to the HA container.

This all used to work very well until a few months ago. Nothing I my setup has changed in a year or so, but sometime over the summer the server was being dropped.

When I look at the configuration in the mobile app after I add the server back, it looks right to me.

Connected via: Internal URL
Version: 2024.11.1
WebSocket: Connected
Local Push: Disabled  

Internal URL: https://ha.blah.network   --- not the real DNS name
External URL: Home Assistant Cloud

When I click into the Internal UR settingsL, I see the SSID of my local network listed correctly and "Local Push" is enabled there.

So it all looks like it should once it's added. Then at some point, I guarantee I'll open the app and it will be gone.

**Comment by bgoncal** [comment]
> > @ChristopherGerdes yes you can find out what is happening because if you export your companion app logs there should be something there to help. Feel free to submit here as well https://forms.gle/Uoqz127Phx4mMTpS6
> 
> Unfortunately, I just reconnected the server. Will I still be able to export this log file from the companion app if it is in new user mode where it's telling me I need to add a server? If so where do I go to do that in that situation? I can grab the log next time it deletes the server.

I was checking your logs and right away when I open your HA cloud URL I get "403", did you disable remote access?

**Comment by bgoncal** [comment]
@jaredgisin I couldn't find your logs, have you submited it?

**Comment by bgoncal** [comment]
I'm experimenting in this PR with showing a message when this happens: https://github.com/home-assistant/iOS/pull/3171

**Comment by ChristopherGerdes** [comment]
> > > @ChristopherGerdes yes you can find out what is happening because if you export your companion app logs there should be something there to help. Feel free to submit here as well https://forms.gle/Uoqz127Phx4mMTpS6
> > 
> > 
> > Unfortunately, I just reconnected the server. Will I still be able to export this log file from the companion app if it is in new user mode where it's telling me I need to add a server? If so where do I go to do that in that situation? I can grab the log next time it deletes the server.
> 
> I was checking your logs and right away when I open your HA cloud URL I get "403", did you disable remote access?

I access home assistant via reverse proxy and do not use the gateway functionality offered by HA cloud.  Is that what you're asking?

**Comment by ChristopherGerdes** [comment]
> I'm experimenting in this PR with showing a message when this happens: #3171

In the past I've gotten a message similar to this when my lets encrypt cert expires (I think).  It is an annoyance but eventually it gets renewed and starts working again.  Is it possible that the app has changed how it reacts to this situation and that's what's causing this?

**Comment by ChristopherGerdes** [comment]
> @ChristopherGerdes got it, weirdly in your logs it shows it trying to connect to HA cloud, can you send me a screenshot of how your external url is configured in the app? of course your can cover any private info

Yup, I think this is what you're wanting:
![image](https://github.com/user-attachments/assets/4309692f-0108-4262-a60a-8cbff64ba392)


Incidentally, is there a way to get to these settings without killing the app and doing the race to hit the config icon before it disappears?  It usually takes me 5 or more tries to get into the menu.

**Comment by bgoncal** [comment]
@ChristopherGerdes the message you saw was probably a red background message saying your SSL certificate was invalid, this message still exists and it is unrelated to this issue as far as I can tell

**Comment by bgoncal** [comment]
@ChristopherGerdes swipe up with 3 fingers

**Comment by ChristopherGerdes** [comment]
> @ChristopherGerdes the message you saw was probably a red background message saying your SSL certificate was invalid, this message still exists and it is unrelated to this issue as far as I can tell

Gotcha, and ya that sounds like the message.  was just sharing in case it helped.

**Comment by ChristopherGerdes** [comment]
> @ChristopherGerdes swipe up with 3 fingers to access server switcher/access to settings

ohh, this is super helpful, thank you!

**Comment by bgoncal** [comment]
New TestFlight beta available

**Comment by jaredgisin** [comment]
> New TestFlight beta available

So far so good on the new beta.
